### PR TITLE
StackedGraphCast

### PR DESCRIPTION
**Pull request by timothyas**
Implements single step GraphCast including casting to half precision, dealing with normalization, and incrementing the initial conditions within InputsAndResiduals.

* main bits are in stacked_graphcast
* stacked_casting has the casting to half precision (not too complicated)
* stacked_normalization has the normalization and deals with incrementing initial conditions. This was tricky, and requires a mapping to get the indices right so that we grab the right input variables and add those to the prediction

It's maybe a bit convoluted, but the implementation can be seen in the GraphUFS repo.

Multi-timestep rollouts are future work since this already got a bit hairy... but shouldn't be too bad given that we already have the input -> target mapping for the InputsAndResiduals. 

Note that this can use the same weights files as the regular graphcast implementation. However, the loss function value is slightly different since the GraphCast loss is averaged per variable, then all the loss per variable is summed up, whereas here, we just take the global (weighted) mean. I've reconstructed the two though, and we can get the same thing, up to some precision. See other PR for that.

### CODE REVIEW

**Comment by danielabdi-noaa** [comment]
@timothyas Nice work! It will take me some time to understand and test your changes here and the graphcast repo too, but rest assured I will get there! Quick question: I am trying to install pytorch/torchvision but it got stuck for hours while solving the environment with the following message
```
Solving environment: | warning  libmamba Added empty dependency for problem type SOLVER_RULE_UPDATE
```
I will try to installing everything freshly from `gpu.yml`. The problem maybe that I had used `gpu-workaround.yml` in the past. If the regular cpu/gpu yml files work now everywhere, maybe we should remove them now?

File: graphcast/casting.py [hunk 1]
```diff
@@ -140,7 +140,7 @@ def _all_inputs_to_bfloat16(
                xarray.Dataset,
                xarray.Dataset]:
   return (inputs.astype(jnp.bfloat16),
-          jax.tree_map(lambda x: x.astype(jnp.bfloat16), targets),
+          jax.tree.map(lambda x: x.astype(jnp.bfloat16), targets),
```
Comment (danielabdi-noaa) [inline graphcast/casting.py]: According to the warning message for the regression testing, `jax.tree.map` requires a higher JAX version while `jax.tree_util.tree_map` works with any version of JAx, so maybe we can use the latter here?

File: graphcast/losses.py [hunk 2]
```diff
@@ -52,6 +57,37 @@ def __call__(self,
...
+    preserves final channel dimension
+
+    Returns:
+        loss_per_sample (chex.Array): total loss per sample within batch
+        loss_per_sample_channel (chex.Array): total loss per channel and per sample
+
+    """
+    # handle potential broadcasting to batch dimension
+    if predictions.ndim == 3:
+        latlon = (0, 1)
```
Comment (danielabdi-noaa) [inline graphcast/losses.py:75]: I suppose this is for the case where the batch dimension is missing.

File: graphcast/stacked_casting.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+        with bfloat16_variable_view():
+            predictions = self._predictor(
+                inputs.astype(jnp.bfloat16),
+                **kwargs,
+            )
+
+        predictions_dtype = infer_floating_dtype(predictions)  # pytype: disable=wrong-arg-types
+        if predictions_dtype != jnp.bfloat16:
+            raise ValueError(f'Expected bfloat16 output, got {predictions_dtype}')
```
Comment (danielabdi-noaa) [inline graphcast/stacked_casting.py:54]: For my understanding: Since with bfloat16 input, we expect the output to be of type bfloat16, I suppose some wrapper (probably the graphcast wrapper in `training.py`) does that? I thought it was meant for casting graphcast  layer weights, not just inputs/outputs, for mixed precision training?

File: graphcast/stacked_casting.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    inputs: chex.Array,
+    targets: chex.Array,
+    weights: Optional[chex.Array | None] = None,
+    ) -> Tuple[chex.Array,
+               chex.Array,
+               chex.Array]:
+
+    i16 = inputs.astype(jnp.bfloat16)
+    t16 = jax.tree.map(lambda x: x.astype(jnp.bfloat16), targets)
+    w16 = weights.astype(jnp.bfloat16) if weights is not None else None
```
Comment (danielabdi-noaa) [inline graphcast/stacked_casting.py:148]: Nevermind we do cast the weights to bfloat16 here.

File: graphcast/stacked_graphcast.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    def __init__(
+        self,
+        model_config: ModelConfig,
+        task_config: TaskConfig
+        ):
+        super().__init__(model_config=model_config, task_config=task_config)
+
+        # since we don't use xarray DataArrays as inputs, we have to
+        # establish the grid somehow. Seems easiest to pass it via task_config
+        # just like the pressure levels
```
Comment (danielabdi-noaa) [inline graphcast/stacked_graphcast.py:24]: Ok this is what you were talking about yesterday about the need to pass lat/lon

File: graphcast/stacked_normalization.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+class StackedInputsAndResiduals(StackedPredictor):
+    """Wraps with a residual connection, normalizing inputs and target residuals.
+
+    The inner predictor is given inputs that are normalized using `locations`
+    and `scales` to roughly zero-mean unit variance.
+
+    For target variables that are present in the inputs, the inner predictor is
+    trained to predict residuals (target - last_frame_of_input) that have been
+    normalized using `residual_scales` (and optionally `residual_locations`) to
+    roughly unit variance / zero mean.
```
Comment (danielabdi-noaa) [inline graphcast/stacked_normalization.py:51]: I didn't know graphcast was also trained to predict residuals, thought that was a Stormer thing.

File: graphcast/stacked_normalization.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    that's based on the scales of the residuals.
+
+    Since we return the underlying predictor's loss on the normalized residuals,
+    if the underlying predictor is a sum of per-variable losses, the normalization
+    will affect the relative weighting of the per-variable loss terms (hopefully
+    in a good way).
+
+    For target variables *not* present in the inputs, the inner predictor is
+    trained to predict targets directly, that have been normalized in the same
+    way as the inputs.
```
Comment (danielabdi-noaa) [inline graphcast/stacked_normalization.py:63]: Ok this is good to know.

**Review by danielabdi-noaa — APPROVED** [review]
@timothyas Nice work! Wish I had your understanding of graphcast code to make such intrusive changes! I have left a few questions but it looks good to me so approving.

File: graphcast/losses.py (hunk 2 above)
Comment (timothyas) [reply graphcast/losses.py:75]: Yeah exactly. I wanted to make StackedGraphCast so that it will handle a squeezed sample (no batch/sample dimension)

File: graphcast/stacked_casting.py (hunk 3 above)
Comment (timothyas) [reply graphcast/stacked_casting.py:54]: I'm actually not sure about this but I'll comment more below.

File: graphcast/stacked_graphcast.py (hunk 5 above)
Comment (timothyas) [reply graphcast/stacked_graphcast.py:24]: Yup exactly. Hey look I did write comments :) (I didn't remember I had done that)

File: graphcast/stacked_normalization.py (hunk 6 above)
Comment (timothyas) [reply graphcast/stacked_normalization.py:51]: Yeah it actually took me a while to understand this, it didn't exactly pop out of the GraphCast paper, and I think the Stormer paper misrepresents GraphCast in that way too. I'm glad I wasn't alone on this.

File: graphcast/stacked_casting.py (hunk 4 above)
Comment (timothyas) [reply graphcast/stacked_casting.py:148]: These are the weights for the loss function, not the neural network weights. DeepMind indicated that the real memory hit is due to the system state, not as much the actual neural network, so I think the NN weights are single precision while the inputs/outputs/targets are half precision.

File: graphcast/casting.py (hunk 1 above)
Comment (timothyas) [reply graphcast/casting.py]: Ah I see, I was mixing those up. I agree, I'll go to tree_util.tree_map. Done in [6058ace](https://github.com/NOAA-PSL/graphcast/pull/5/commits/6058ace9939187413d48eb9aa68ee40f3df472e6)

File: graphcast/stacked_casting.py (hunk 4 above)
Comment (timothyas) [reply graphcast/stacked_casting.py:148]: Yeah running `print( jax.tree_util.tree_map( lambda x : x.dtype, params ))` confirms that the neural net weights are all float32 even though the system state is bfloat16

**Comment by timothyas** [comment]
Thanks for reviewing this @danielabdi-noaa. I'll go ahead and merge this PR.
